Error while checking build/O2/fullCI for f949cd79526a819585af61b16fb032634d6f36a5 at 2024-06-27 21:55:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/13254-slc8_x86-64/0/PWGCF/GenericFramework/Tasks/flowGenericFramework.cxx:498:17: error: 'const struct o2::soa::Table<o2::aod::mcparticle::Phi, o2::aod::mcparticle::Eta, o2::aod::mcparticle::Pt, o2::aod::mcparticle::P, o2::aod::mcparticle::Y, o2::soa::Index<0, -1>, o2::aod::mcparticle::McCollisionId, o2::aod::mcparticle::PdgCode, o2::aod::mcparticle::StatusCode, o2::aod::mcparticle::Flags, o2::aod::mcparticle::MothersIds, o2::aod::mcparticle::DaughtersIdSlice, o2::aod::mcparticle::Weight, o2::aod::mcparticle::Px, o2::aod::mcparticle::Py, o2::aod::mcparticle::Pz, o2::aod::mcparticle::E, o2::aod::mcparticle::Vx, o2::aod::mcparticle::Vy, o2::aod::mcparticle::Vz, o2::aod::mcparticle::Vt, o2::aod::mcparticle::PVector<o2::aod::mcparticle::Px, o2::aod::mcparticle::Py, o2::aod::mcparticle::Pz>, o2::aod::mcparticle::ProducedByGenerator<o2::aod::mcparticle::Flags>, o2::aod::mcparticle::FromBackgroundEvent<o2::aod::mcparticle::Flags>, o2::aod::mcparticle::GetGenStatusCode<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::GetHepMCStatusCode<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::GetProcess<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::IsPhysicalPrimary<o2::aod::mcparticle::Flags> >::RowViewBase<o2::soa::DefaultIndexPolicy, o2::soa::Join<o2::aod::McParticles_001Extension, o2::soa::Table<o2::soa::Index<0, -1>, o2::aod::mcparticle::McCollisionId, o2::aod::mcparticle::PdgCode, o2::aod::mcparticle::StatusCode, o2::aod::mcparticle::Flags, o2::aod::mcparticle::MothersIds, o2::aod::mcparticle::DaughtersIdSlice, o2::aod::mcparticle::Weight, o2::aod::mcparticle::Px, o2::aod::mcparticle::Py, o2::aod::mcparticle::Pz, o2::aod::mcparticle::E, o2::aod::mcparticle::Vx, o2::aod::mcparticle::Vy, o2::aod::mcparticle::Vz, o2::aod::mcparticle::Vt, o2::aod::mcparticle::PVector<o2::aod::mcparticle::Px, o2::aod::mcparticle::Py, o2::aod::mcparticle::Pz>, o2::aod::mcparticle::ProducedByGenerator<o2::aod::mcparticle::Flags>, o2::aod::mcparticle::FromBackgroundEvent<o2::aod::mcparticle::Flags>, o2::aod::mcparticle::GetGenStatusCode<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::GetHepMCStatusCode<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::GetProcess<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::IsPhysicalPrimary<o2::aod::mcparticle::Flags> > >, o2::aod::McParticles_001Extension, o2::soa::Table<o2::soa::Index<0, -1>, o2::aod::mcparticle::McCollisionId, o2::aod::mcparticle::PdgCode, o2::aod::mcparticle::StatusCode, o2::aod::mcparticle::Flags, o2::aod::mcparticle::MothersIds, o2::aod::mcparticle::DaughtersIdSlice, o2::aod::mcparticle::Weight, o2::aod::mcparticle::Px, o2::aod::mcparticle::Py, o2::aod::mcparticle::Pz, o2::aod::mcparticle::E, o2::aod::mcparticle::Vx, o2::aod::mcparticle::Vy, o2::aod::mcparticle::Vz, o2::aod::mcparticle::Vt, o2::aod::mcparticle::PVector<o2::aod::mcparticle::Px, o2::aod::mcparticle::Py, o2::aod::mcparticle::Pz>, o2::aod::mcparticle::ProducedByGenerator<o2::aod::mcparticle::Flags>, o2::aod::mcparticle::FromBackgroundEvent<o2::aod::mcparticle::Flags>, o2::aod::mcparticle::GetGenStatusCode<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::GetHepMCStatusCode<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::GetProcess<o2::aod::mcparticle::Flags, o2::aod::mcparticle::StatusCode>, o2::aod::mcparticle::IsPhysicalPrimary<o2::aod::mcparticle::Flags> > >' has no member named 'tpcNClsFound'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13254/f949cd79526a819585af61b16fb032634d6f36a5/build_O2_fullCI/pretty.html).

--- a/Framework/Core/include/Framework/ASoA.h
+++ b/Framework/Core/include/Framework/ASoA.h
@@ -1466,7 +1466,7 @@ class Table
     auto getId() const
     {
       using decayed = std::decay_t<TI>;
-      if constexpr (framework::has_type_v<decayed, bindings_pack_t>) { // index to another table
+      if constexpr (framework::has_type<decayed>(bindings_pack_t{})) { // index to another table
         constexpr auto idx = framework::has_type_at_v<decayed>(bindings_pack_t{});
         return framework::pack_element_t<idx, external_index_columns_t>::getId();
       } else if constexpr (std::is_same_v<decayed, Parent>) { // self index
@@ -2777,12 +2777,12 @@ struct Join : TableWrap<Ts...>::table_t {
   }
 
   template <typename T>
-  static constexpr bool contains()
+  static consteval bool contains()
   {
     if constexpr (is_type_with_originals_v<T>) {
       return contains(typename T::originals{});
     } else {
-      return framework::has_type_v<T, originals>;
+      return framework::has_type<T>(originals{});
     }
   }
 

--- a/Framework/Core/include/Framework/AnalysisHelpers.h
+++ b/Framework/Core/include/Framework/AnalysisHelpers.h
@@ -90,7 +90,7 @@ struct WritingCursor<soa::Table<PC...>> {
     if constexpr (soa::is_soa_iterator_v<T>) {
       return arg.globalIndex();
     } else {
-      static_assert(!framework::has_type_v<T, framework::pack<PC...>>, "Argument type mismatch");
+      static_assert(!framework::has_type<T>(framework::pack<PC...>{}), "Argument type mismatch");
       return arg;
     }
   }

--- a/Framework/Core/include/Framework/AnalysisManagers.h
+++ b/Framework/Core/include/Framework/AnalysisManagers.h
@@ -50,7 +50,7 @@ struct GroupedCombinationManager<GroupedCombinationsGenerator<T1, GroupingPolicy
   {
     static_assert(sizeof...(T2s) > 0, "There must be associated tables in process() for a correct pair");
     if constexpr (std::is_same_v<G, TG>) {
-      static_assert(std::conjunction_v<framework::has_type<As, pack<T2s...>>...>, "You didn't subscribed to all tables requested for mixing");
+      static_assert((framework::has_type<As>(pack<T2s...>{}) && ...), "You didn't subscribed to all tables requested for mixing");
       comb.setTables(grouping, associated);
     }
   }

--- a/Framework/Core/include/Framework/BinningPolicy.h
+++ b/Framework/Core/include/Framework/BinningPolicy.h
@@ -240,7 +240,7 @@ struct FlexibleBinningPolicy<std::tuple<Ls...>, Ts...> : BinningPolicyBase<sizeo
   template <typename T, typename T2>
   auto getBinningValue(T& rowIterator, arrow::Table* table, uint64_t ci = -1, uint64_t ai = -1, uint64_t globalIndex = -1) const
   {
-    if constexpr (has_type_v<T2, pack<Ls...>>) {
+    if constexpr (has_type<T2>(pack<Ls...>{})) {
       if (globalIndex != -1) {
         rowIterator.setCursor(globalIndex);
       }

--- a/Framework/Foundation/include/Framework/Pack.h
+++ b/Framework/Foundation/include/Framework/Pack.h
@@ -175,16 +175,14 @@ void print_pack()
 template <template <typename> typename Condition, typename... Types>
 using filtered_pack = std::decay_t<decltype(filter_pack<Condition>(pack<>{}, pack<Types...>{}))>;
 
-/// Check if a given pack Pack has a type T inside.
-template <typename T, typename Pack>
-struct has_type;
-
 template <typename T, typename... Us>
-struct has_type<T, pack<Us...>> : std::disjunction<std::is_same<T, Us>...> {
-};
+bool consteval has_type(framework::pack<Us...>)
+{
+  return (std::is_same_v<T, Us> || ...);
+}
 
-template <typename T, typename... Us>
-inline constexpr bool has_type_v = has_type<T, Us...>::value;
+template <typename T, typename P>
+inline constexpr bool has_type_v = has_type<T>(P{});
 
 template <template <typename, typename> typename Condition, typename T, typename Pack>
 struct has_type_conditional;
@@ -207,7 +205,7 @@ constexpr size_t has_type_at(pack<T1, Ts...> const&)
 {
   if constexpr (std::is_same_v<T, T1>) {
     return 0;
-  } else if constexpr (has_type_v<T, pack<Ts...>>) {
+  } else if constexpr (has_type<T>(pack<Ts...>{})) {
     return 1 + has_type_at<T>(pack<Ts...>{});
   }
   return sizeof...(Ts) + 2;
@@ -274,7 +272,7 @@ struct intersect_pack {
   {
     return filtered_pack<std::is_void,
                          std::conditional_t<
-                           has_type_v<pack_element_t<Indices, S1>, S2>,
+                           has_type<pack_element_t<Indices, S1>>(S2{}),
                            pack_element_t<Indices, S1>, void>...>{};
   }
   using type = decltype(make_intersection(std::make_index_sequence<pack_size(S1{})>{}));
@@ -306,7 +304,7 @@ struct subtract_pack {
   {
     return filtered_pack<std::is_void,
                          std::conditional_t<
-                           !has_type_v<pack_element_t<Indices, S1>, S2>,
+                           !has_type<pack_element_t<Indices, S1>>(S2{}),
                            pack_element_t<Indices, S1>, void>...>{};
   }
   using type = decltype(make_subtraction(std::make_index_sequence<pack_size(S1{})>{}));

--- a/Framework/Foundation/test/test_FunctionalHelpers.cxx
+++ b/Framework/Foundation/test/test_FunctionalHelpers.cxx
@@ -28,6 +28,8 @@ struct TestStruct {
 TEST_CASE("TestOverride")
 {
   static_assert(pack_size(pack<int, float>{}) == 2, "Bad size for the pack");
+  static_assert(has_type<int>(pack<int, float>{}) == true, "int should be in the pack");
+  static_assert(has_type<double>(pack<int, float>{}) == false, "double should not be in the pack");
   static_assert(has_type_v<int, pack<int, float>> == true, "int should be in the pack");
   static_assert(has_type_v<double, pack<int, float>> == false, "double should not be in the pack");
   static_assert(has_type_conditional_v<std::is_same, int, pack<int, float>> == true, "int should be in the pack");
@@ -43,7 +45,7 @@ TEST_CASE("TestOverride")
   static_assert(std::is_same_v<selected_pack_multicondition<is_same_as_second_t, pack<int>, pack<int, float, char>>, pack<int>>, "multiselector should select int");
   static_assert(std::is_same_v<filtered_pack<is_int_t, int, float, char>, pack<float, char>>, "filter should remove int");
   static_assert(std::is_same_v<intersected_pack_t<pack<int, float, char>, pack<float, double>>, pack<float>>, "filter intersect two packs");
-  static_assert(has_type_v<pack_element_t<0, pack<int>>, pack<int>> == true, "foo");
+  static_assert(has_type<pack_element_t<0, pack<int>>>(pack<int>{}) == true, "foo");
   print_pack<intersected_pack_t<pack<int>, pack<int>>>();
   print_pack<intersected_pack_t<pack<TestStruct<0, -1>, int>, pack<TestStruct<0, -1>, float>>>();
   static_assert(std::is_same_v<intersected_pack_t<pack<TestStruct<0, -1>, int>, pack<TestStruct<0, -1>, float>>, pack<TestStruct<0, -1>>>, "filter intersect two packs");

